### PR TITLE
added lucene sentence query argument to apis

### LINKED ISSUE #5: Add ability for a filtering query in the API
The basic idea is to add another query to the API, similar to the parentQuery that will also allow to filter the results based on some additional metadata available on the lucene documents. 

Because some of the metadata we might want is not available in the Index, we also want a simple way to expand the indexing process by adding more metadata to sentences.

The basic design is to provide hooks in odinson in two places:
1. At indexing time, to allow the addition of more metadata to the index
2. At search time to supplement the searchQuery with additional constraints (based on metadata in the index) - this will ustilize the new query added to the API as input.

In addition we can group the two corresponding hooks as a single plugin and we should be able to  define a list (or hierarchy) of plugins to load using configuration.
  
Open questions:
1. When using multiple plugins how exactly will they interact? Are they aware of each other or are they not? How do we split the query from the API to send only the relevant parts of the query to every plugin. 
2. What should happen if the index does not contain the metadata expected by a plugin? For example if there was a mismatch where the index was created without the plugin but is trying to run a query with it. Can we somehow store this metadata (what plugins were used in creating the index) in the index itself so that this can be validated when odinson back end is getting up?

### PR DESCRIPTION
**Pull request by mcshlain**
Part of: https://github.com/allenai/spike/issues/328

This PR adds:

-  to both `/search` and `/rich-search` apis a new property called `sentenceQuery` for an additional arbitrary lucene query. 
- The lucene query can be provided by itself without the `odinsonQuery`
- The lucene query defaults to `word` so for example `2007 tag:NN` is the same as `word:2007 tag:NN`
- It's possible to write complex queries with lucene syntax (see https://docs.microsoft.com/en-us/azure/search/query-lucene-syntax#bkmk_boolean)
   - Examples:
      - `founded -1999` - with word `founded` but without the word `1999`.
      - `founded || founder` - with word `founded` or with the word `founder`.

### CODE REVIEW

File: core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala [hunk 1]
```diff
@@ -81,11 +83,11 @@ class ExtractorEngine(
 
   /** executes query and returns next n results after the provided doc */
   def query(
-    odinsonQuery: String,
-    n: Int,
-    afterDoc: Int,
-    afterScore: Float
-  ): OdinResults = {
+             odinsonQuery: String,
```
Comment (yaarams) [inline core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala]: change idea's tab formatting to 2 spaces and save to fix this indendation

File: core/src/main/scala/ai/lum/odinson/lucene/search/OdinsonIndexSearcher.scala [hunk 2]
```diff
@@ -44,4 +49,43 @@ class OdinsonIndexSearcher(
     search(query, manager)
   }
 
+  def pureSearch(query: Query, n: Int): OdinResults = {
+
+    // Modify query to only return sentence entries
+    val builder = new BooleanQuery.Builder()
```
Comment (yaarams) [inline core/src/main/scala/ai/lum/odinson/lucene/search/OdinsonIndexSearcher.scala:63]: add tests for the pureSearch:
- doesnt retrieve type: parent
- continuation works
- positive flow works (simpoe/or query)

File: core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala [hunk 3]
```diff
@@ -150,6 +152,47 @@ class ExtractorEngine(
...
+      case (None, Some(sq)) => sq // TODO: convert to some kind of odinson query
+      case (Some(oq), None) => oq
+      case _ => throw new IllegalArgumentException("Either odinson query or a sentence query have to be defined")
+    }
+
+    val numberOfDocuments = extractionQuery.numDocuments.getOrElse(numDocs())
+
+    // run the query
+    (fullQuery, extractionQuery.continuation) match {
+      case (q: OdinsonQuery, Some(c)) => indexSearcher.odinSearch(c, q, numberOfDocuments)
```
Comment (yaarams) [inline core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala]: call all the options, and check that the response makes sense:
- odinson, odinson+search, search, continuation

File: core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala (hunk 1 above)
Comment (mcshlain) [reply core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala]: It wasn't tabs just different formatting style (in this case for multi line function definition). Added scalafmt to the project (code style format) and formatted the files with it, the function definitions are now as they were before but it also changed other things but as long as its standard and not incidental, I'm ok with it.

File: backend/public/schema/odinson.yaml [hunk 4]
```diff
@@ -1,825 +1,841 @@
...
+      in: query
+      type: string
+      description: An Odison pattern.
+      example: "[lemma=phosphorylate] []"
+
+    SentenceQuery:
+      name: sentenceQuery
+      in: query
+      type: string
+      description: A lucene query on a sentence.
```
Comment (yaarams) [inline backend/public/schema/odinson.yaml]: lucene => Lucene

Comment (mcshlain) [reply backend/public/schema/odinson.yaml]: fixed

File: core/src/main/scala/ai/lum/odinson/lucene/search/OdinsonIndexSearcher.scala (hunk 2 above)
Comment (mcshlain) [reply core/src/main/scala/ai/lum/odinson/lucene/search/OdinsonIndexSearcher.scala:63]: tests added in `ai.lum.odinson.lucene.search.OdinsonIndexSearcherTest`

File: core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala (hunk 3 above)
Comment (mcshlain) [reply core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala]: tests added in `ai.lum.odinson.ExtractorEngineTest`

**Review by yaarams — APPROVED** [review]

**Pull request merged**